### ISSUE TITLE
[Question] Action Cfg Discrete Space

### ISSUE THREAD

**Issue by shiven2001** [isaac-sim/IsaacLab#2795]
So it seems that there are no examples (at least I didn't see it, if there are please let me know) that use discrete action space in IsaacLab. My task is following:

I have a floating camera and I want the action space to be pose and orientation change in each dt meaning the camera can move in x,y,z by 1 unit or decide to rotate along x,y,z by 10 degrees. I made the code according to what I thought should be ok but it seems to have a couple of issues:

      from enum import Enum, auto
      from typing import Sequence, Tuple, Type
      import os
      import sys
      import torch
      from pydantic import BaseModel
      import math
      
      from isaaclab.utils import configclass
      from isaaclab.managers import ActionTerm, ActionTermCfg
      from isaaclab.assets import RigidObject
      from isaaclab.envs import ManagerBasedEnv
      
      script_dir = os.path.dirname(os.path.abspath(__file__))
      sys.path.append(script_dir)
      
      from xutils.xmath import deg_to_rad
      from xutils.xlogging import logging
      
      
      class FloatingCameraActionDiscrete(ActionTerm):
          cfg: "FloatingCameraActionCfg"
          _env: ManagerBasedEnv
          _asset: RigidObject
      
          def __init__(self, cfg: "FloatingCameraActionCfg", env: ManagerBasedEnv):
              super().__init__(cfg, env)
      
              # Initialize asset
              self._asset = env.scene[cfg.asset_name]
              self._initial_state = self._asset.data.root_state_w  # Shape: (num_envs, 13)
      
              # Setup action space
              self._setup_action_space()
      
              # Initialize action tensors
              self._raw_actions = torch.zeros((self.num_envs, 1), device=self.device)
              self._processed_actions = torch.zeros(
                  (self.num_envs, 6), device=self.device
              )  # pos (3) + rot (3)
      
              # Initialize current state (position + quaternion)
              self._current_state = self._initial_state.clone()  # Shape: (num_envs, 13)
              # Set initial position to (0, 0, 1.5) relative to env origin, keep initial quaternion
              self._current_state[:, :3] = torch.tensor([0.0, 0.0, 1.5], device=self.device)
      
              # Define position bounds (relative to env origin)
              self._pos_bounds_min = torch.tensor(
                  [
                      self.cfg.pos_bounds[0][0],
                      self.cfg.pos_bounds[1][0],
                      self.cfg.pos_bounds[2][0],
                  ],
                  device=self.device,
              )
              self._pos_bounds_max = torch.tensor(
                  [
                      self.cfg.pos_bounds[0][1],
                      self.cfg.pos_bounds[1][1],
                      self.cfg.pos_bounds[2][1],
                  ],
                  device=self.device,
              )
      
              # Define orientation bounds (radians, from cfg.rot_bounds)
              self._rot_bounds_min = torch.tensor(
                  [
                      deg_to_rad(self.cfg.rot_bounds[0][0]),
                      deg_to_rad(self.cfg.rot_bounds[1][0]),
                      deg_to_rad(self.cfg.rot_bounds[2][0]),
                  ],
                  device=self.device,
              )
              self._rot_bounds_max = torch.tensor(
                  [
                      deg_to_rad(self.cfg.rot_bounds[0][1]),
                      deg_to_rad(self.cfg.rot_bounds[1][1]),
                      deg_to_rad(self.cfg.rot_bounds[2][1]),
                  ],
                  device=self.device,
              )
      
              # Define discrete action map: 0=stay, 1-6=position (±x, ±y, ±z), 7-12=rotation (±x, ±y, ±z)
              self._action_map = {
                  0: torch.tensor([0.0, 0.0, 0.0, 0.0, 0.0, 0.0], device=self.device),  # Stay
                  1: torch.tensor(
                      [self.cfg.spacing, 0.0, 0.0, 0.0, 0.0, 0.0], device=self.device
                  ),  # +x
                  2: torch.tensor(
                      [-self.cfg.spacing, 0.0, 0.0, 0.0, 0.0, 0.0], device=self.device
                  ),  # -x
                  3: torch.tensor(
                      [0.0, self.cfg.spacing, 0.0, 0.0, 0.0, 0.0], device=self.device
                  ),  # +y
                  4: torch.tensor(
                      [0.0, -self.cfg.spacing, 0.0, 0.0, 0.0, 0.0], device=self.device
                  ),  # -y
                  5: torch.tensor(
                      [0.0, 0.0, self.cfg.spacing, 0.0, 0.0, 0.0], device=self.device
                  ),  # +z
                  6: torch.tensor(
                      [0.0, 0.0, -self.cfg.spacing, 0.0, 0.0, 0.0], device=self.device
                  ),  # -z
                  7: torch.tensor(
                      [0.0, 0.0, 0.0, deg_to_rad(self.cfg.rot_angle), 0.0, 0.0],
                      device=self.device,
                  ),  # +rot_x
                  8: torch.tensor(
                      [0.0, 0.0, 0.0, -deg_to_rad(self.cfg.rot_angle), 0.0, 0.0],
                      device=self.device,
                  ),  # -rot_x
                  9: torch.tensor(
                      [0.0, 0.0, 0.0, 0.0, deg_to_rad(self.cfg.rot_angle), 0.0],
                      device=self.device,
                  ),  # +rot_y
                  10: torch.tensor(
                      [0.0, 0.0, 0.0, 0.0, -deg_to_rad(self.cfg.rot_angle), 0.0],
                      device=self.device,
                  ),  # -rot_y
                  11: torch.tensor(
                      [0.0, 0.0, 0.0, 0.0, 0.0, deg_to_rad(self.cfg.rot_angle)],
                      device=self.device,
                  ),  # +rot_z
                  12: torch.tensor(
                      [0.0, 0.0, 0.0, 0.0, 0.0, -deg_to_rad(self.cfg.rot_angle)],
                      device=self.device,
                  ),  # -rot_z
              }
      
          @property
          def action_dim(self) -> int:
              return 1  # Single discrete action selecting from 13 choices
      
          @property
          def raw_actions(self) -> torch.Tensor:
              return self._raw_actions
      
          @property
          def processed_actions(self) -> torch.Tensor:
              return self._processed_actions
      
          def process_actions(self, actions: torch.Tensor):
              """Process discrete actions into position and rotation increments, clipping both."""
              self._raw_actions = actions
              self._processed_actions = torch.zeros((self.num_envs, 6), device=self.device)
      
              # Validate and clip actions to valid range [0, 12]
              actions = actions.clone().to(dtype=torch.long, device=self.device)
              actions = torch.clamp(actions, 0, 12)  # Ensure actions are within valid range
      
              # Map discrete actions to position and rotation increments
              for i in range(self.num_envs):
                  action_idx = int(actions[i].item())
                  if action_idx in self._action_map:
                      self._processed_actions[i] = self._action_map[action_idx]
                  else:
                      logging.warning(
                          f"Invalid action {action_idx} for env {i} after clamping, staying in place. "
                          f"Original action: {actions[i].item()}"
                      )
                      self._processed_actions[i] = self._action_map[0]
      
              # Extract position and rotation increments
              pos_inc = self._processed_actions[:, :3]  # (num_envs, 3)
              rot_inc = self._processed_actions[:, 3:6]  # (num_envs, 3)
      
              # Update positions with clipping
              new_positions = self._current_state[:, :3] + pos_inc
              new_positions = torch.clamp(
                  new_positions, self._pos_bounds_min, self._pos_bounds_max
              )
      
              # Update orientations with clipping
              current_quat = self._current_state[:, 3:7]  # (num_envs, 4)
              new_quat = current_quat.clone()
              for i in range(self.num_envs):
                  if torch.any(rot_inc[i] != 0):
                      # Convert current quaternion to Euler angles
                      euler = self._quaternion_to_euler(current_quat[i])  # (roll, pitch, yaw)
                      # Apply rotation increment
                      new_euler = euler + rot_inc[i]
                      # Clip Euler angles to configured bounds
                      new_euler = torch.clamp(
                          new_euler, self._rot_bounds_min, self._rot_bounds_max
                      )
                      # Convert back to quaternion
                      new_quat[i] = self._euler_to_quaternion(new_euler)
      
              # Update current state
              self._current_state[:, :3] = new_positions
              self._current_state[:, 3:7] = new_quat
      
          def apply_actions(self):
              """Apply processed actions to set camera position and orientation."""
              # Use current state for positions and quaternions
              positions = self._current_state[:, :3]  # (num_envs, 3)
              quats = self._current_state[:, 3:7]  # (num_envs, 4)
      
              # Transform to world frame by adding env origins
              env_origins = self._env.scene.env_origins  # (num_envs, 3)
              world_positions = positions + env_origins
      
              # Zero velocities
              linear_vel = torch.zeros((self.num_envs, 3), device=self.device)
              angular_vel = torch.zeros((self.num_envs, 3), device=self.device)
      
              # Combine into root state: [x, y, z, qw, qx, qy, qz, vx, vy, vz, wx, wy, wz]
              root_state = torch.cat(
                  [world_positions, quats, linear_vel, angular_vel], dim=-1
              )
      
              # Apply to asset
              env_ids = torch.arange(self.num_envs, device=self.device)
              self._asset.write_root_state_to_sim(root_state, env_ids=env_ids)
      
          def reset(self, env_ids: Sequence[int] | None = None):
              """Reset camera to initial state and zero actions."""
              if env_ids is None:
                  env_ids = torch.arange(self.num_envs, device=self.device)
              else:
                  env_ids = torch.tensor(env_ids, device=self.device)
      
              # Reset actions
              self._raw_actions[env_ids] = 0.0
              self._processed_actions[env_ids] = 0.0
      
              # Reset to initial state: position (0, 0, 1.5), initial quaternion
              self._current_state[env_ids, :3] = torch.tensor(
                  [0.0, 0.0, 1.5], device=self.device
              )
              self._current_state[env_ids, 3:7] = self._initial_state[env_ids, 3:7]
      
              # Apply reset pose
              self.apply_actions()
      
          def _setup_action_space(self):
              """Configure the discrete action space."""
              logging.debug(
                  f"Camera action space: Discrete with 13 actions (stay, ±x, ±y, ±z, ±rot_x, ±rot_y, ±rot_z), "
                  f"position step={self.cfg.spacing}m, rotation step={self.cfg.rot_angle}deg, "
                  f"position clipped to x [{self.cfg.pos_bounds[0][0]}, {self.cfg.pos_bounds[0][1]}], "
                  f"y [{self.cfg.pos_bounds[1][0]}, {self.cfg.pos_bounds[1][1]}], "
                  f"z [{self.cfg.pos_bounds[2][0]}, {self.cfg.pos_bounds[2][1]}], "
                  f"rotation clipped to roll [{self.cfg.rot_bounds[0][0]}, {self.cfg.rot_bounds[0][1]}]deg, "
                  f"pitch [{self.cfg.rot_bounds[1][0]}, {self.cfg.rot_bounds[1][1]}]deg, "
                  f"yaw [{self.cfg.rot_bounds[2][0]}, {self.cfg.rot_bounds[2][1]}]deg"
              )
      
          def _euler_to_quaternion(self, euler: torch.Tensor) -> torch.Tensor:
              """Convert Euler angles (radians) to quaternion (qw, qx, qy, qz)."""
              roll, pitch, yaw = euler
              cr = torch.cos(roll / 2)
              sr = torch.sin(roll / 2)
              cp = torch.cos(pitch / 2)
              sp = torch.sin(pitch / 2)
              cy = torch.cos(yaw / 2)
              sy = torch.sin(yaw / 2)
              qw = cr * cp * cy + sr * sp * sy
              qx = sr * cp * cy - cr * sp * sy
              qy = cr * sp * cy + sr * cp * sy
              qz = cr * cp * sy - sr * sp * cy
              return torch.tensor([qw, qx, qy, qz], device=self.device)
      
          def _quaternion_to_euler(self, quat: torch.Tensor) -> torch.Tensor:
              """Convert quaternion (qw, qx, qy, qz) to Euler angles (roll, pitch, yaw) in radians."""
              w, x, y, z = quat
              sinr_cosp = 2 * (w * x + y * z)
              cosr_cosp = 1 - 2 * (x * x + y * y)
              roll = torch.atan2(sinr_cosp, cosr_cosp)
              sinp = 2 * (w * y - z * x)
              pitch = torch.asin(torch.clamp(sinp, -1.0, 1.0))
              siny_cosp = 2 * (w * z + x * y)
              cosy_cosp = 1 - 2 * (y * y + z * z)
              yaw = torch.atan2(siny_cosp, cosy_cosp)
              return torch.tensor([roll, pitch, yaw], device=self.device)
      
      
      class CameraControlMode(Enum):
          DISCRETE = auto()
          CONTINUOUS = auto()
      
      
      class CameraConfig(BaseModel):
          position: Tuple[float, float, float]
      
      
      @configclass
      class FloatingCameraActionDiscreteCfg(ActionTermCfg):
          class_type: Type = FloatingCameraActionDiscrete
          asset_name: str = "float_cam"
          scale: float = 1.0
          control_mode: CameraControlMode = CameraControlMode.DISCRETE
          pos_bounds: Tuple[Tuple[float, float], Tuple[float, float], Tuple[float, float]] = (
              (-10.0, 10.0),  # x
              (-10.0, 10.0),  # y
              (0.0, 10.0),  # z
          )
          rot_bounds: Tuple[Tuple[float, float], Tuple[float, float], Tuple[float, float]] = (
              (-90.0, 90.0),  # roll
              (-90.0, 90.0),  # pitch
              (-90.0, 90.0),  # yaw
          )
          spacing: float = 1.0
          rot_angle: float = 10.0
      

ISSUE 1: I always test my stuff is working using the random_agent.py file but when I tried this discrete space, the camera didn't move at all. 

ISSUE 2: I tried with the rsl_rl train.py with PPO default config and then camera moves it seems it is not working because most of the time, camera stays at a place or maybe moves left or right a bit along x axis but or rotation and movement along other axis happens. I am quite sure it is my bad implementation of the discrete space. Can sometime tell me where I went wrong and how to fix this? I think an example of discrete space would really benefit the community.

**Comment by Toni-SM** [comment]
Hi @shiven2001 

Refer to the Isaac Lab's [Spaces showcase](https://isaac-sim.github.io/IsaacLab/main/source/overview/environments.html#spaces-showcase) documentation for examples on defining/using the various Gymnasium observation and action spaces supported in Isaac Lab.

Also, see the https://github.com/isaac-sim/IsaacLab/tree/main/source/isaaclab_tasks/isaaclab_tasks/direct/cartpole_showcase code, particularly in your case:

https://github.com/isaac-sim/IsaacLab/blob/66cc743e24ec3a99274d2e5cc285d8f584e2506a/source/isaaclab_tasks/isaaclab_tasks/direct/cartpole_showcase/cartpole/cartpole_env_cfg.py#L47-L72

https://github.com/isaac-sim/IsaacLab/blob/66cc743e24ec3a99274d2e5cc285d8f584e2506a/source/isaaclab_tasks/isaaclab_tasks/direct/cartpole_showcase/cartpole/cartpole_env.py#L26-L29

**Comment by shiven2001** [comment]
it says that:

"Currently, only Isaac Lab’s Direct workflow supports the definition of observation and action spaces other than Box. See Direct workflow’s [observation_space](https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.envs.html#isaaclab.envs.DirectRLEnvCfg.observation_space) / [action_space](https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.envs.html#isaaclab.envs.DirectRLEnvCfg.action_space) documentation for more details."

Does that mean I cannot use discrete space in Manager Based Env?

**Comment by Toni-SM** [comment]
Hi @shiven2001 

> Does that mean I cannot use discrete space in Manager Based Env?

Only the Isaac Lab’s Direct workflow supports the different spaces
